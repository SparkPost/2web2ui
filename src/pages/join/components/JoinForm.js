import React, { Component } from 'react';
import { connect } from 'react-redux';
import { reduxForm, Field, getFormValues } from 'redux-form';
import Recaptcha from 'react-recaptcha';

import config from 'src/config';
import { FORMS, LINKS } from 'src/constants';
import { TextFieldWrapper, CheckboxWrapper } from 'src/components/reduxFormWrappers';
import { Form } from 'src/components/form';
import { ExternalLink } from 'src/components/links';
import { Box, Button, Checkbox, Stack } from 'src/components/matchbox';
import { required, minLength, email, endsWithWhitespace } from 'src/helpers/validation';
const { recaptcha } = config;

/** Recaptcha flow
 *
 * UX:
 * Disable form until recaptcha is ready
 * Challenge with recaptcha when Create Account button is clicked
 * Submits form data to api upon successful completetion of recaptcha challenge


 * Code
 * Recaptcha script is loaded in JoinPage
 * When script is loaded, Recaptcha component will initialize it and bind to different callbacks
 * ref: reference to recaptcha instance
 * onloadCallback: it's triggered upon Recaptcha's dependencies (global) is completely loaded. we enable form at that point
 * executeRecaptcha: this is triggered, when user clicks Create Account button. captcha challenge is presented to user and submitted to Google for validation
 * verifyCallback: this is triggered upon user completes recaptcha challenge successfuly. we pass the data we've received
 */

export class JoinForm extends Component {
  state = {
    reCaptchaReady: false,
    reCaptchaInstance: null,
    recaptcha_response: null,
    showPassword: false,
  };

  reCaptchaLoaded = () => {
    this.setState({ reCaptchaReady: true });
  };

  onSubmit = formValues => {
    this.props.onSubmit({
      ...formValues,
      recaptcha_response: this.state.recaptcha_response,
      recaptcha_type: 'invisible',
    });
  };

  recaptchaVerifyCallback = response => {
    this.state.reCaptchaInstance.reset();
    this.setState({ recaptcha_response: response }, this.props.handleSubmit(this.onSubmit));
  };

  executeRecaptcha = event => {
    event.preventDefault();
    this.state.reCaptchaInstance.execute();
  };

  linkRecaptcha = instance => {
    if (!this.state.reCaptchaInstance) {
      this.setState({ reCaptchaInstance: instance });
    }
  };

  render() {
    const { loading, pristine, invalid, submitting } = this.props;

    const { reCaptchaReady, showPassword } = this.state;

    const pending = loading || submitting || !reCaptchaReady;

    return (
      <Form id="sign-up-form" onSubmit={this.executeRecaptcha}>
        <Stack>
          <Field
            name="first_name"
            component={TextFieldWrapper}
            label="First Name"
            autoComplete="given-name"
            validate={required}
            disabled={pending}
          />
          <Field
            name="last_name"
            component={TextFieldWrapper}
            label="Last Name"
            autoComplete="family-name"
            validate={required}
            disabled={pending}
          />
          <Field
            name="company_name"
            component={TextFieldWrapper}
            label="Company"
            disabled={pending}
            autoComplete="organization"
          />
          <Field
            name="email"
            component={TextFieldWrapper}
            label="Email"
            validate={[required, email]}
            disabled={pending}
            autoComplete="username email"
          />
          <Field
            name="password"
            component={TextFieldWrapper}
            label="Password"
            validate={[required, minLength(12), endsWithWhitespace]}
            disabled={!reCaptchaReady || loading}
            type={showPassword ? 'text' : 'password'}
            autoComplete="new-password"
            connectRight={
              <Button
                variant="connected"
                data-id="show-password-button"
                onClick={() => this.setState({ showPassword: !showPassword })}
              >
                {showPassword ? 'Hide' : 'Show'}
              </Button>
            }
          />

          <Checkbox.Group>
            <Field
              name="email_opt_in"
              id="email_opt_in"
              component={CheckboxWrapper}
              disabled={pending}
              type="checkbox"
              label={<span>I'm happy to receive marketing email from SparkPost</span>}
            />

            <Field
              name="tou_accepted"
              id="tou_accepted"
              component={CheckboxWrapper}
              type="checkbox"
              disabled={pending}
              validate={required}
              label={
                <span>
                  I agree to SparkPost's <ExternalLink to={LINKS.TOU}>Terms of Use</ExternalLink>
                </span>
              }
            />
          </Checkbox.Group>

          <Box>
            <Button
              id="submit"
              type="submit"
              variant="primary"
              disabled={pending || pristine || invalid}
            >
              {loading ? 'Loading' : 'Create Account'}
            </Button>
          </Box>
        </Stack>

        <Recaptcha
          ref={this.linkRecaptcha}
          sitekey={recaptcha.invisibleKey}
          size="invisible"
          render="explicit"
          onloadCallback={this.reCaptchaLoaded}
          verifyCallback={this.recaptchaVerifyCallback}
        />
      </Form>
    );
  }
}

const mapStateToProps = state => ({
  initialValues: {
    tou_accepted: false,
    email_opt_in: false,
  },
  loading: state.account.createLoading || state.auth.loginPending,
  formValues: getFormValues(FORMS.JOIN)(state),
});

const RegisterAccountForm = reduxForm({ form: FORMS.JOIN })(JoinForm);
export default connect(mapStateToProps, {})(RegisterAccountForm);
