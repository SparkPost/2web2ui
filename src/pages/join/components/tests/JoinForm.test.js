import { shallow } from 'enzyme';
import React from 'react';
import { JoinForm } from '../JoinForm';

let props;
let instance;
let wrapper;

describe('JoinForm', () => {
  let mockRecaptcha;

  beforeEach(() => {
    mockRecaptcha = {
      execute: jest.fn(),
      reset: jest.fn(),
    };

    props = {
      loading: false,
      onSubmit: jest.fn(),
      handleSubmit: jest.fn(),
    };

    wrapper = shallow(<JoinForm {...props} />);
    instance = wrapper.instance();
    wrapper.setState({ reCaptchaInstance: mockRecaptcha });
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  describe('render', () => {
    it('disables form when loading', () => {
      wrapper.setProps({ loading: true });
      expect(wrapper).toMatchSnapshot();
    });

    it('disables form when recaptcha is not loaded', () => {
      wrapper.setState({ reCaptchaReady: false });
      expect(wrapper).toMatchSnapshot();
    });

    it('enables form correctly', () => {
      wrapper.setProps({ loading: false, pristine: false, invalid: false });
      wrapper.setState({ reCaptchaReady: true });
      expect(wrapper).toMatchSnapshot();
    });

    describe('submit', () => {
      it('invokes recaptcha challenge on click', () => {
        const form = wrapper.find('Form').first();
        form.simulate('submit', { preventDefault: jest.fn() });
        expect(mockRecaptcha.execute).toHaveBeenCalledTimes(1);
      });
    });

    describe('executeRecaptcha', () => {
      it('invokes recaptcha on click of submit button', () => {
        wrapper.setState({ reCaptchaInstance: mockRecaptcha });
        instance.executeRecaptcha({ preventDefault: jest.fn() });
        expect(mockRecaptcha.execute).toHaveBeenCalledTimes(1);
      });
    });

    describe('reCaptchaLoaded', () => {
      it('sets reCaptchaReady to be true', () => {
        instance.reCaptchaLoaded();
        expect(wrapper.state().reCaptchaReady).toBe(true);
      });
    });

    describe('linkRecaptcha', () => {
      it('sets reCaptchaInstance to state if not set', () => {
        instance.setState({ reCaptchaInstance: null });
        instance.linkRecaptcha({});
        expect(wrapper.state().reCaptchaInstance).toEqual({});
      });

      it('does not set reCaptchaInstance to state if already set', () => {
        instance.linkRecaptcha({});
        expect(wrapper.state().reCaptchaInstance).toEqual(mockRecaptcha);
      });
    });

    describe('recaptchaVerifyCallback', () => {
      it('resets recaptcha', () => {
        instance.recaptchaVerifyCallback('foobar');
        expect(mockRecaptcha.reset).toHaveBeenCalledTimes(1);
      });

      it('sets recaptcha response to state & invokes handleSubmit', () => {
        const formValues = {
          first_name: 'foo',
          last_name: 'bar',
          email: 'foo@bar.com',
          password: '***',
        };
        wrapper.setProps({ formValues });
        instance.recaptchaVerifyCallback('foobar');
        expect(wrapper.state().recaptcha_response).toEqual('foobar');
        expect(props.handleSubmit).toHaveBeenCalledWith(instance.onSubmit);
      });
    });

    describe('onSubmit', () => {
      beforeEach(() => {
        wrapper.setState({ recaptcha_response: 'some-response' });
      });

      it('submits with correct data', () => {
        const formValues = {
          first_name: 'foo',
          last_name: 'bar',
          email: 'foo@bar.com',
          password: '***',
        };
        instance.onSubmit(formValues);
        expect(props.onSubmit).toHaveBeenCalledWith({
          ...formValues,
          recaptcha_response: 'some-response',
          recaptcha_type: 'invisible',
        });
      });
    });
  });
});
