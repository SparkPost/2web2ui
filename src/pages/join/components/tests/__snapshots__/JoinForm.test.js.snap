// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JoinForm render disables form when loading 1`] = `
<Form
  id="sign-up-form"
>
  <Stack>
    <Field
      autoComplete="given-name"
      component={[Function]}
      disabled={true}
      label="First Name"
      name="first_name"
      validate={[Function]}
    />
    <Field
      autoComplete="family-name"
      component={[Function]}
      disabled={true}
      label="Last Name"
      name="last_name"
      validate={[Function]}
    />
    <Field
      autoComplete="organization"
      component={[Function]}
      disabled={true}
      label="Company"
      name="company_name"
    />
    <Field
      autoComplete="username email"
      component={[Function]}
      disabled={true}
      label="Email"
      name="email"
      validate={
        Array [
          [Function],
          [Function],
        ]
      }
    />
    <Field
      autoComplete="new-password"
      component={[Function]}
      connectRight={
        <Button
          data-id="show-password-button"
          onClick={[Function]}
          variant="connected"
        >
          Show
        </Button>
      }
      disabled={true}
      label="Password"
      name="password"
      type="password"
      validate={
        Array [
          [Function],
          [Function],
          [Function],
        ]
      }
    />
    <Checkbox.Group>
      <Field
        component={[Function]}
        disabled={true}
        id="email_opt_in"
        label={
          <span>
            I'm happy to receive marketing email from SparkPost
          </span>
        }
        name="email_opt_in"
        type="checkbox"
      />
      <Field
        component={[Function]}
        disabled={true}
        id="tou_accepted"
        label={
          <span>
            I agree to SparkPost's 
            <ExternalLink
              to="https://www.sparkpost.com/policies/tou"
            >
              Terms of Use
            </ExternalLink>
          </span>
        }
        name="tou_accepted"
        type="checkbox"
        validate={[Function]}
      />
    </Checkbox.Group>
    <Box>
      <Button
        disabled={true}
        id="submit"
        onClick={[Function]}
        variant="primary"
      >
        Loading
      </Button>
    </Box>
  </Stack>
  <t
    badge="bottomright"
    elementID="g-recaptcha"
    expiredCallbackName="expiredCallback"
    hl="en"
    onloadCallback={[Function]}
    onloadCallbackName="onloadCallback"
    render="explicit"
    sitekey="6LekChoUAAAAAJZouMPHnhRss2t7-ZetbAABfsOZ"
    size="invisible"
    tabindex="0"
    theme="light"
    type="image"
    verifyCallback={[Function]}
    verifyCallbackName="verifyCallback"
  />
</Form>
`;

exports[`JoinForm render disables form when recaptcha is not loaded 1`] = `
<Form
  id="sign-up-form"
>
  <Stack>
    <Field
      autoComplete="given-name"
      component={[Function]}
      disabled={true}
      label="First Name"
      name="first_name"
      validate={[Function]}
    />
    <Field
      autoComplete="family-name"
      component={[Function]}
      disabled={true}
      label="Last Name"
      name="last_name"
      validate={[Function]}
    />
    <Field
      autoComplete="organization"
      component={[Function]}
      disabled={true}
      label="Company"
      name="company_name"
    />
    <Field
      autoComplete="username email"
      component={[Function]}
      disabled={true}
      label="Email"
      name="email"
      validate={
        Array [
          [Function],
          [Function],
        ]
      }
    />
    <Field
      autoComplete="new-password"
      component={[Function]}
      connectRight={
        <Button
          data-id="show-password-button"
          onClick={[Function]}
          variant="connected"
        >
          Show
        </Button>
      }
      disabled={true}
      label="Password"
      name="password"
      type="password"
      validate={
        Array [
          [Function],
          [Function],
          [Function],
        ]
      }
    />
    <Checkbox.Group>
      <Field
        component={[Function]}
        disabled={true}
        id="email_opt_in"
        label={
          <span>
            I'm happy to receive marketing email from SparkPost
          </span>
        }
        name="email_opt_in"
        type="checkbox"
      />
      <Field
        component={[Function]}
        disabled={true}
        id="tou_accepted"
        label={
          <span>
            I agree to SparkPost's 
            <ExternalLink
              to="https://www.sparkpost.com/policies/tou"
            >
              Terms of Use
            </ExternalLink>
          </span>
        }
        name="tou_accepted"
        type="checkbox"
        validate={[Function]}
      />
    </Checkbox.Group>
    <Box>
      <Button
        disabled={true}
        id="submit"
        onClick={[Function]}
        variant="primary"
      >
        Create Account
      </Button>
    </Box>
  </Stack>
  <t
    badge="bottomright"
    elementID="g-recaptcha"
    expiredCallbackName="expiredCallback"
    hl="en"
    onloadCallback={[Function]}
    onloadCallbackName="onloadCallback"
    render="explicit"
    sitekey="6LekChoUAAAAAJZouMPHnhRss2t7-ZetbAABfsOZ"
    size="invisible"
    tabindex="0"
    theme="light"
    type="image"
    verifyCallback={[Function]}
    verifyCallbackName="verifyCallback"
  />
</Form>
`;

exports[`JoinForm render enables form correctly 1`] = `
<Form
  id="sign-up-form"
>
  <Stack>
    <Field
      autoComplete="given-name"
      component={[Function]}
      disabled={false}
      label="First Name"
      name="first_name"
      validate={[Function]}
    />
    <Field
      autoComplete="family-name"
      component={[Function]}
      disabled={false}
      label="Last Name"
      name="last_name"
      validate={[Function]}
    />
    <Field
      autoComplete="organization"
      component={[Function]}
      disabled={false}
      label="Company"
      name="company_name"
    />
    <Field
      autoComplete="username email"
      component={[Function]}
      disabled={false}
      label="Email"
      name="email"
      validate={
        Array [
          [Function],
          [Function],
        ]
      }
    />
    <Field
      autoComplete="new-password"
      component={[Function]}
      connectRight={
        <Button
          data-id="show-password-button"
          onClick={[Function]}
          variant="connected"
        >
          Show
        </Button>
      }
      disabled={false}
      label="Password"
      name="password"
      type="password"
      validate={
        Array [
          [Function],
          [Function],
          [Function],
        ]
      }
    />
    <Checkbox.Group>
      <Field
        component={[Function]}
        disabled={false}
        id="email_opt_in"
        label={
          <span>
            I'm happy to receive marketing email from SparkPost
          </span>
        }
        name="email_opt_in"
        type="checkbox"
      />
      <Field
        component={[Function]}
        disabled={false}
        id="tou_accepted"
        label={
          <span>
            I agree to SparkPost's 
            <ExternalLink
              to="https://www.sparkpost.com/policies/tou"
            >
              Terms of Use
            </ExternalLink>
          </span>
        }
        name="tou_accepted"
        type="checkbox"
        validate={[Function]}
      />
    </Checkbox.Group>
    <Box>
      <Button
        disabled={false}
        id="submit"
        onClick={[Function]}
        variant="primary"
      >
        Create Account
      </Button>
    </Box>
  </Stack>
  <t
    badge="bottomright"
    elementID="g-recaptcha"
    expiredCallbackName="expiredCallback"
    hl="en"
    onloadCallback={[Function]}
    onloadCallbackName="onloadCallback"
    render="explicit"
    sitekey="6LekChoUAAAAAJZouMPHnhRss2t7-ZetbAABfsOZ"
    size="invisible"
    tabindex="0"
    theme="light"
    type="image"
    verifyCallback={[Function]}
    verifyCallbackName="verifyCallback"
  />
</Form>
`;
